### ISSUE TITLE
cli config 

### ISSUE THREAD

**Issue by davidszotten** [onefinestay/nameko#223]
could we let extensions register argparse options (e.g. via setuptools entrypoints) for configuration when run via `nameko run`?

could we extend this to a nicer way to set options when hosting via code (i.e. something better than importing `FOO_CONFIG_KEY`)

**Comment by kollhof** [comment]
How about accepting a config-file param?
As the number of extensions grows the list of args would grow making it a bit painful to list them as command line args. 
To manage different sets of settings one is probably more likely to keep them in files, too.

**Comment by mattbennett** [comment]
Commandeering this issue to capture the discussion about registering configuration options.

Thanks to https://github.com/onefinestay/nameko/pull/255 we now have the ability to load config from a file, but there are a few other issues:

* ``nameko run`` accepts a ``--broker`` option but none of the other configuration keys
* most config is used by Extensions, which can be separate projects, but there's no way to inspect which config values are required or what their defaults are
* importing ``FOO_CONFIG_KEY`` everywhere is ugly and you have to know where each Extension chooses to define them

If we registered configuration options and default values (via setuptools entrypoints?) we could:

a) access them from a known location (e.g. ``container.config``)
b) inspect the required keys and any default values for a given set of extensions
c) automatically populate the ``argparse`` options for ``nameko run`` as @davidszotten suggests

**Comment by rochacbruno** [comment]
@mattbennett what about reading environment variables?

    export NAMEKO_AMQP_URI="amqp://guest:guest@localhost"

then in nameko.cli.main

    import os
    config = {k, v for k, v in os.environ if k.startswith('NAMEKO')}

**Comment by mattbennett** [comment]
@rochacbruno yep, I think env vars should be supported, but as part of a larger piece of work.

**Comment by fabiocerqueira** [comment]
Maybe this example could be a good way to use environment variables on YAML config file: 
http://stackoverflow.com/questions/26712003/pyyaml-parsing-of-the-environment-variable-in-the-yaml-configuration-file

**Comment by fabiocerqueira** [comment]
Maybe this example could be a good way to use environment variables on YAML config file: 
http://stackoverflow.com/questions/26712003/pyyaml-parsing-of-the-environment-variable-in-the-yaml-configuration-file

**Comment by fabiocerqueira** [comment]
```python
ENVVAR_PATTERN = re.compile(r'\$\{([a-zA-Z][^}:\s]+)(:([^}]+))?\}')
yaml.add_implicit_resolver ("!envvar", ENVVAR_PATTERN)


def replace_envvar(match):
    env_var, _, default = match.groups()
    if default == 'null':
        default = None
    return os.environ.get(env_var, default)


def envvar_constructor(loader, node):
    value = loader.construct_scalar(node)
    line = ENVVAR_PATTERN.sub(replace_envvar, value)
    return line
yaml.add_constructor('!envvar', envvar_constructor)
```

**Comment by davidszotten** [comment]
this looks really nice. i like the bash-style `${VAR_NAME<:DEFAULT_VALUE>}`. @fabiocerqueira  would you like to raise a pull-request that implements this?

**Comment by kooba** [comment]
@davidszotten  @fabiocerqueira I'll be happy to pick this up if that's ok.

**Comment by fabiocerqueira** [comment]
@kooba go ahead :D I thinks this would be great. It will be nice if supports different types, int values, float values, boolean values. Thank you.

**Comment by rejoc** [comment]
It would be interesting to allow multiple config files.

In many applications, you have part of the configuration that is common to many/all services (like AMQP_URI, exchange, serializer, database url...) and some services need vry specific paramters.

Allowing --config to be a list (nargs='+') and updating config with these files would do the job 

something like this:
```
config = {}
for conf_file in args.config):
    with open(conf_file) as fle:
            config.update(yaml.load(fle))
```

**Comment by mattbennett** [comment]
@rejoc I like this idea. We could bundle it into #520 (cc @iky) or make it as a standalone change first. Do you want to raise a PR?

**Comment by davidszotten** [comment]
is it common for tools to support multiple config files like this? (none of the ones i can think of do)

(i tend to handle this kind of sharing in my config management tooling)
